File: src/orchestrator/registry.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::path::PathBuf;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InstanceEntry {
    pub slug: String,
    pub config_path: String,
    pub state_dir: String,
    pub started_at: DateTime<Utc>,
}

#[derive(Debug, Default, Serialize, Deserialize)]
pub struct InstanceRegistry {
    pub instances: Vec<InstanceEntry>,
}

impl InstanceRegistry {
    fn registry_path() -> PathBuf {
        crate::platform::home_dir()
            .unwrap_or_else(|| PathBuf::from("/tmp"))
            .join(".devrig")
            .join("instances.json")
    }

    pub fn load() -> Self {
        let path = Self::registry_path();
        if let Ok(content) = std::fs::read_to_string(&path) {
            serde_json::from_str(&content).unwrap_or_default()
        } else {
            Self::default()
        }
    }

    pub fn save(&self) -> anyhow::Result<()> {
        let path = Self::registry_path();
        if let Some(parent) = path.parent() {
            std::fs::create_dir_all(parent)?;
        }
        let content = serde_json::to_string_pretty(self)?;
        let tmp_path = path.with_extension("json.tmp");
        std::fs::write(&tmp_path, &content)?;
        std::fs::rename(&tmp_path, &path)?;
        Ok(())
    }

    pub fn register(&mut self, entry: InstanceEntry) {
        // Update existing entry or add new one
        if let Some(existing) = self.instances.iter_mut().find(|e| e.slug == entry.slug) {
            *existing = entry;
        } else {
            self.instances.push(entry);
        }
    }

    pub fn unregister(&mut self, slug: &str) {
        self.instances.retain(|e| e.slug != slug);
    }

    pub fn list(&self) -> &[InstanceEntry] {
        &self.instances
    }

    /// Remove entries whose state files no longer exist
    pub fn cleanup(&mut self) {
        self.instances.retain(|entry| {
            let state_path = PathBuf::from(&entry.state_dir).join("state.json");
            state_path.exists()
        });
    }
}

File: internal/registry/registry.go
// Package registry maintains the global list of running devrig instances so
// that commands like `devrig ps` and `devrig stop` can find projects without
// a config file argument.
package registry

import (
	"encoding/json"
	"os"
	"path/filepath"
	"time"
)

// InstanceEntry records a running (or recently stopped) devrig project.
type InstanceEntry struct {
	Slug       string    `json:"slug"`
	ConfigPath string    `json:"config_path"`
	StateDir   string    `json:"state_dir"`
	StartedAt  time.Time `json:"started_at"`
}

// Registry is the list stored at ~/.devrig/instances.json.
type Registry struct {
	Instances []InstanceEntry `json:"instances"`
}

func registryPath() string {
	home, err := os.UserHomeDir()
	if err != nil {
		home = "/tmp"
	}
	return filepath.Join(home, ".devrig", "instances.json")
}

// Load reads instances.json; returns an empty registry if the file doesn't exist.
func Load() *Registry {
	data, err := os.ReadFile(registryPath())
	if err != nil {
		return &Registry{}
	}
	var r Registry
	if err := json.Unmarshal(data, &r); err != nil {
		return &Registry{}
	}
	return &r
}

// Save writes instances.json atomically (tmp → rename).
func (r *Registry) Save() error {
	path := registryPath()
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	data, err := json.MarshalIndent(r, "", "  ")
	if err != nil {
		return err
	}
	tmp := path + ".tmp"
	if err := os.WriteFile(tmp, data, 0o644); err != nil {
		return err
	}
	return os.Rename(tmp, path)
}

// Register adds or updates an entry for the given slug.
func (r *Registry) Register(entry InstanceEntry) {
	for i, e := range r.Instances {
		if e.Slug == entry.Slug {
			r.Instances[i] = entry
			return
		}
	}
	r.Instances = append(r.Instances, entry)
}

// Unregister removes the entry with the given slug.
func (r *Registry) Unregister(slug string) {
	keep := r.Instances[:0]
	for _, e := range r.Instances {
		if e.Slug != slug {
			keep = append(keep, e)
		}
	}
	r.Instances = keep
}

// Cleanup removes entries whose state.json no longer exists.
func (r *Registry) Cleanup() {
	keep := r.Instances[:0]
	for _, e := range r.Instances {
		if _, err := os.Stat(filepath.Join(e.StateDir, "state.json")); err == nil {
			keep = append(keep, e)
		}
	}
	r.Instances = keep
}
